### DISCUSSION TITLE
Different RQL generated for query in 5.x versus 4.x

### DISCUSSION THREAD

**Discussion by mmoore99** [ravendb/ravendb#16378] [Q&A] [answered]
When executing the following query, the RavenDb client generates different RQL in 4.x versus 5.x when a document being used in a query is referenced in a `CustomContractResolver` which overrides the `JsonProperty` attribute for properties by using `property.UnderlyingName`.

```
var startDate = new DateTime(2023, 05, 01);
var endDate = new DateTime(2023, 05, 01);
var orders = _dbLocal.Query<ShopifyOrder>().Where(order => order.CreatedAt.Date >= startDate.Date && order.CreatedAt.Date <= endDate.Date).ToList();
```

The RQL generated for the 4.x query:
`from ShopifyOrders where CreatedAt.Date between $p0 and $p1 {"p0":"2023-05-01T00:00:00.0000000","p1":"2023-05-01T00:00:00.0000000"}`

The RQL generated for the  5.x query:
`from 'ShopifyOrders' where created_at.Date between $p0 and $p1 {"p0":"2023-05-01T00:00:00.0000000","p1":"2023-05-01T00:00:00.0000000"}`

Note "**CreatedAt.Date**" versus "**created_at.Date**". It appears that the `CustomContractResolver` (see below) is not being taken into consideration when generating the property name for the RQL used in the query in RavenDb 5.x. 

**This worked correctly in 4.x. Have I done something wrong in the migration or is this a bug?**

The document being queried looks like this:
```
public class ShopifyOrder
{
    [JsonProperty("id")]
    public string Id { get; set; }

    [JsonProperty("email")]
    public string Email { get; set; }

    [JsonProperty("closed_at")]
    public DateTimeOffset? ClosedAt { get; set; }

    [JsonProperty("created_at")]
    public DateTimeOffset CreatedAt { get; set; }
}
```

I have the following initialization code in place which had to be changed in the migration to 5.x.

```
_documentStore.Conventions.IdentityPartsSeparator = '-';
var serializationConventions = new NewtonsoftJsonSerializationConventions
{
     CustomizeJsonSerializer = x => { x.ObjectCreationHandling = ObjectCreationHandling.Replace; },
 };
 serializationConventions.JsonContractResolver = new CustomContractResolver(serializationConventions);
 _documentStore.Conventions.Serialization = serializationConventions;
```

The CustomContractResolver is as follows:

```
public class CustomContractResolver : DefaultRavenContractResolver
    {
        public CustomContractResolver(ISerializationConventions conventions) : base(conventions)
        {
        }

        public static bool CanWrite(MemberInfo memberInfo)
        {
            switch (memberInfo)
            {
                case PropertyInfo _:
                    return ((PropertyInfo)memberInfo).CanWrite;
                case FieldInfo _:
                    return true;
            }

            throw new NotSupportedException("Cannot calculate CanWrite on " + memberInfo);
        }

        // If enabled, this code effectively causes the [JsonProperty] attribute to be ignored during serialization of specific types
        // which allow the property names for those types to be serialized using the case as in the C# object
        protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
        {
            // Let the base class create all the properties. If [JsonProperty] attribute exists the name in the attribute will be assigned to the property
            var properties = base.CreateProperties(type, memberSerialization);

            if (type.FullName == typeof(ShopifyProduct).FullName 
                || (type.DeclaringType != null && type.DeclaringType.FullName == typeof(ShopifyProduct).FullName) 
                
                || type.FullName == typeof(ShopifyOrder).FullName 
                || (type.DeclaringType != null && type.DeclaringType.FullName == typeof(ShopifyOrder).FullName))
            {
                // Now inspect each property and replace the name with the C# property name which is PascalCase
                foreach (var property in properties)
                {
                    property.PropertyName = property.UnderlyingName;
                }
            }

            return properties;
        }

    }

```

**Comment by garayx** [comment]
hi,

Can you prepare a failing unit test or a simple hello world app, that reproduces the issue, so we can take a look on it?

https://ravendb.net/docs/article-page/5.4/Csharp/start/test-driver#complete-example

**Comment by mmoore99** [comment]
The link below contains a solution with a failing unit test for the issue as described above using RavenDb 5.4. Please let me know if you need anything else and advise me of your findings.

https://www.dropbox.com/t/VvVYTfPIkV1S16dS

**Reply by garayx** [reply]
Thanks, we will check it out, and let you know if any additional info is needed

**Comment by Lwiel** [comment]
What is the exact version it used to work on? I checked test you provided on 4.2.123, and the behavior is the same as for versions 5.x.

I would also like to know what do you want to achieve by using `CustomContractResolver` here.
Currently you serialize properties under lowercased names, but store documents with C# fieldnames. Effictively you ignore the fieldname set in `JsonProperty` attribute, while this attribute is respected by querying.

**Comment by mmoore99** [comment]
> What is the exact version it used to work on?

The production client version that has been working for a long time is 4.0.3.

I attempted to setup a test using the TestRunner but I couldn't get it working with the 4.0.3 client so I have created a console app using the live test server at http://live-test.ravendb.net/. 

The link to the solution for the console app is https://www.dropbox.com/t/oP4L0tOMaEhKYGOT

You will notice the console app has some differences from the original test that I sent you. The queries in the new console app more completely show which queries worked and did not work using the 4.0.3 client.

> I would also like to know what do you want to achieve by using CustomContractResolver here.

The source for the entity in question is an external api. The `JsonProperty` attribute is there to allow for the incoming json api data to be deserialized into the C# field names. The `CustomContractResolver` allows the data stored in RavenDB to use the C# field names. Since the query using `CreatedAt.Date` had previously always worked I did not know that the `JsonProperty` attribute was respected when querying.

**Reply by Lwiel** [reply]
Query with `CreatedAt.Date` worked by accident on RavenDB version 4.0.3. There was a bug where `JsonProperty` name was not respected for nested paths when generating RQL. It was fixed in later version.

For your case, it would probably be the best to have two separate entities - one for dealing with api data (with `JsonProperty` names included), and one for RavenDB (without `JsonProperty` names). You can use tools such as [AutoMapper](https://github.com/AutoMapper/AutoMapper) to do this automatically.

**Reply by rachkoud** [reply]
> Currently you serialize properties under lowercased names, but store documents with C# fieldnames. Effictively you ignore the fieldname set in JsonProperty attribute, **while this attribute is respected by querying.**

> There was a bug where JsonProperty name was not respected for nested paths when generating RQL. It was fixed in later version.

@Lwiel I had this issue recently and I'm interested to know how can we have the same behaviour from storing/quering by ignoring JsonProperty, so nested path ignore also JsonProperty when generating RQL ? I didn't found in the doc where we can configure RQL generation serialization behaviour : 

[Conventions: Serialization](https://ravendb.net/docs/article-page/6.0/csharp/client-api/configuration/serialization)

**Reply by garayx** [reply]
I guess using a different set of entities (without JsonProperty) for storing/querying RavenDB would work for this case.
But I wondering why would you set a `JsonProperty` in 1st place ?

**Reply by rachkoud** [reply]
@garayx the entities are coming from a NuGet package with many nested properties and many classes. They are used by an API client, it's really easier to save the entities as they are defined, it works well for storying the entities (JsonProperty is ignored) but for querying it doesn't work (JsonProperty is not ignored). So if we can just configure an option to ignore JsonProperty with RQL generation that would be nice.

**Reply by ppekrol** [reply]
Something that comes to my mind is to take leverage from one of those:
- DocumentConventions.FindPropertyNameForIndex
- DocumentConventions.FindPropertyNameForDynamicIndex
- or use DocumentQuery?

Using the 'FindPropertyNameForDynamicIndex' you can generate from:

```csharp
DocumentStore.Conventions.FindPropertyNameForDynamicIndex = (indexedType, indexedName, path, prop) =>
{
    if (indexedType == typeof(User))
    {
        // your logic here

        return "better_name";
    }

    return DocumentConventions.DefaultFindPropertyNameForDynamicIndex(indexedType, indexedName, path, prop);
};
```

```csharp
session.Query<User>().Where(x => x.Name == "bob").ToList();
```

Following RQL:
```sql
from 'Users' where better_name = $p0
```

**Reply by rachkoud** [reply]
@ppekrol that's very interesting!

My type is [GitCommitRef](https://learn.microsoft.com/en-us/dotnet/api/microsoft.teamfoundation.sourcecontrol.webapi.gitcommitref?view=azure-devops-dotnet) which has a nested property Author.Date : 

[GitCommitRef.Author](https://learn.microsoft.com/en-us/dotnet/api/microsoft.teamfoundation.sourcecontrol.webapi.gitcommitref.author?view=azure-devops-dotnet#microsoft-teamfoundation-sourcecontrol-webapi-gitcommitref-author)
![image](https://github.com/ravendb/ravendb/assets/1620056/e653c0c7-a169-4890-94c3-189567678ed4)

[GitUserDate.Date](https://learn.microsoft.com/en-us/dotnet/api/microsoft.teamfoundation.sourcecontrol.webapi.gituserdate.date?view=azure-devops-dotnet#microsoft-teamfoundation-sourcecontrol-webapi-gituserdate-date)
![image](https://github.com/ravendb/ravendb/assets/1620056/0a3f0a2d-a7fa-4df8-a582-e6f455708a74)

This query : 

```csharp
var commitsQuery = session.Query<Commit>()
    .Where(w =>
        w.Author.Date >= from &&
        w.Author.Date <= to &&
        w.A == username)
    .Take(take)
    .Skip(skip)
    .OrderByDescending(p => p.Author.Date);
```

Generates this RQL : 
 
```sql
from 'Commits' where author.date between $p0 and $p1 and A = $p2 order by author.date desc limit $p3, $p4
```

By using your example : 

```csharp
FindPropertyNameForDynamicIndex = (indexedType, indexedName, path, prop) =>
{
    if (indexedType == typeof(Commit))
    {
        // your logic here

        return "better_name";
    }

    return DocumentConventions.DefaultFindPropertyNameForDynamicIndex(indexedType, indexedName, path, prop);
};
```

I'm getting : 

```sql
from 'Commits' where better_name between $p0 and $p1 and better_name = $p2 order by better_name desc limit $p3, $p4
```

And the query **works fine** (without any results of course) : 

```csharp
await commitsQuery.ToListAsync();
```

But if I just try to fix my issue like this : 

```csharp
FindPropertyNameForDynamicIndex = (indexedType, indexedName, path, prop) =>
{
    if (indexedType == typeof(Commit))
    {
        if (prop == "author.date")
        {
            return "Author.Date";    
        }
    }
    
    return DocumentConventions.DefaultFindPropertyNameForDynamicIndex(indexedType, indexedName, path, prop);
}
```

commitsQuery.ToString(); generates this without any exception (the issue is fixed from the query perspective) : 

```sql
from 'Commits' where Author.Date between $p0 and $p1 and A = $p2
```

But I'm getting this exception by calling `await commitsQuery.ToListAsync()` 

Do you know why?

![image](https://github.com/ravendb/ravendb/assets/1620056/22e33097-1fe2-430e-846c-b821cc5f1ca4)

System.ArgumentException: An item with the same key has already been added. Key: Changes
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Collections.ObjectModel.KeyedCollection`2.AddKey(TKey key, TItem item)
   at System.Collections.ObjectModel.KeyedCollection`2.InsertItem(Int32 index, TItem item)
   at System.Collections.ObjectModel.Collection`1.Add(T item)
   at Newtonsoft.Json.Utilities.CollectionUtils.AddRange[T](IList`1 initial, IEnumerable`1 collection)
   at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateObjectContract(Type objectType)
   at Raven.Client.Json.Serialization.NewtonsoftJson.DefaultRavenContractResolver.CreateObjectContract(Type objectType) in C:\Builds\RavenDB-Stable-5.4\54001\src\Raven.Client\Json\Serialization\NewtonsoftJson\DefaultRavenContractResolver.cs:line 113
   at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateContract(Type objectType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Newtonsoft.Json.Utilities.ThreadSafeStore`2.Get(TKey key)
   at Newtonsoft.Json.Serialization.DefaultContractResolver.ResolveContract(Type type)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.GetContract(Type type)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.GetContractSafe(Type type)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
   at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.NewtonsoftJsonJsonSerializer.Raven.Client.Json.Serialization.IJsonSerializer.Deserialize(IJsonReader reader, Type type) in C:\Builds\RavenDB-Stable-5.4\54001\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\NewtonsoftJsonJsonSerializer.cs:line 10
   at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.NewtonsoftJsonBlittableEntitySerializer.EntityFromJsonStream(Type type, BlittableJsonReaderObject json) in C:\Builds\RavenDB-Stable-5.4\54001\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\NewtonsoftJsonBlittableEntitySerializer.cs:line 64
   at Raven.Client.Json.Serialization.NewtonsoftJson.NewtonsoftJsonSerializationConventions.Raven.Client.Json.Serialization.ISerializationConventions.DeserializeEntityFromBlittable(Type type, BlittableJsonReaderObject json) in C:\Builds\RavenDB-Stable-5.4\54001\src\Raven.Client\Json\Serialization\NewtonsoftJson\NewtonsoftJsonSerializationConventions.cs:line 155
   at Raven.Client.Json.Serialization.NewtonsoftJson.Internal.SessionBlittableJsonConverter.FromBlittable(Type type, BlittableJsonReaderObject& json, String id, Boolean trackEntity) in C:\Builds\RavenDB-Stable-5.4\54001\src\Raven.Client\Json\Serialization\NewtonsoftJson\Internal\SessionBlittableJsonConverter.cs:line 56

**Reply by Lwiel** [reply]
Could you prepare a failing unit test or a console application that reproduces the issue, so we can take a look on it?

https://ravendb.net/docs/article-page/5.4/Csharp/start/test-driver#complete-example

**Reply by rachkoud** [reply, marked as answer]
@Lwiel @ppekrol after recreating a [small console app that you can find here it is working now](https://github.com/rachkoud/ravendb-demos/tree/main/RavenDb.Serizalization.Problem) 👍

First, I use a specific JsonContractResolver to use the UnderlyingName and ignore DataMember or JsonProperty property **when storing a document**

```csharp
public class CustomizedRavenJsonContractResolver : DefaultRavenContractResolver
{
    public CustomizedRavenJsonContractResolver(ISerializationConventions conventions) : base(conventions)
    {
    }

    protected override JsonProperty CreateProperty(MemberInfo member, MemberSerialization memberSerialization)
    {
        var jsonProperty = base.CreateProperty(member, memberSerialization);
        jsonProperty.PropertyName = jsonProperty.UnderlyingName;
        return jsonProperty;
    }
}
```

With this CustomizedRavenJsonContractResolver, for example, theses **DataMember** are ignored : 

```csharp
[DataContract]
[KnownType(typeof (GitCommit))]
public class GitCommitRef : VersionControlSecuredObject
{
  /// <summary>ID (SHA-1) of the commit.</summary>
  [DataMember(Name = "commitId", EmitDefaultValue = false)]
  public string CommitId { get; set; }

  /// <summary>Author of the commit.</summary>
  [DataMember(Name = "author", EmitDefaultValue = false)]
  public GitUserDate Author { get; set; }

  /// <summary>Comment or message of the commit.</summary>
  [DataMember(Name = "comment", EmitDefaultValue = false)]
  public string Comment { get; set; }
```

So it will save it like this : 

```json
{
    "Author": {
        "Email": "author@demo.com",
        "Date": "2023-05-15T14:08:08.7017811"
    },
    "Comment": "Dummy comment",
```

But if we query it like this, we will not get any results, **because RQL generation uses DataMember or JsonProperty and not our CustomizedRavenJsonContractResolver** : 

```csharp
session.Query<GitCommitRef>()
        .Where(w =>
            w.Author.Date >= DateTime.Today.AddDays(-2) &&
            w.Author.Date <= DateTime.Today.AddDays(1) &&
            w.Author.Email == email);
```

It will generate this query :

```sql
from 'GitCommitRefs' where author.date between $p0 and $p1 and author.email = $p2
```

instead of this query : 

```sql
from 'GitCommitRefs' where Author.Date between $p0 and $p1 and Author.Email = $p2
```

So to fix this, we must use this convention to capitalize the first letter : 

```csharp
FindPropertyNameForDynamicIndex = (indexedType, indexedName, path, prop) =>
{
    if (indexedType == typeof(GitCommitRef))
    {
        // Capitalize first letter of the prop 
        // author.email becomes Author.Email
        string[] properties = prop.Split('.');
        for (int i = 0; i < properties.Length; i++)
        {
            if (!string.IsNullOrEmpty(properties[i]))
            {
                properties[i] = char.ToUpper(properties[i][0]) + properties[i].Substring(1);
            }
        }
                
        return string.Join(".", properties);
    }
                
    return DocumentConventions.DefaultFindPropertyNameForDynamicIndex(indexedType, indexedName, path, prop);
}
```

You will find [a working example with the issue and the fix here](https://github.com/rachkoud/ravendb-demos/tree/main/RavenDb.Serizalization.Problem).

Thanks for support @ppekrol 

@mmoore99 it could be interesting for your problem

**Comment by mmoore99** [comment]
@rachkoud thanks for the suggestion, it does appear that may be a fix for my issue.

What is the purpose of `FindIdentityProperty = x => x.Name == "CommitId" || x.Name == "Id",` at Program.cs, line 26?

**Reply by rachkoud** [reply]
@mmoore99 because in my case I don't have an Id column, you can can guide RavenDB with [FindIdentityProperty](https://ravendb.net/docs/article-page/5.4/csharp/client-api/configuration/identifier-generation/global#findidentityproperty) but you probably don't need it.

The client must know where in your entity an identifier is stored to be properly able to transform it into JSON document. It uses the FindIdentityProperty convention for that. The default and very common convention is that a property named Id is the identifier, so is the default one:

```csharp
FindIdentityProperty = memberInfo => memberInfo.Name == "Id"
```

You can provide a customization based on the MemberInfo parameter to indicate which property or field keeps the identifier. The client will iterate over all object properties and take the first one according to the defined predicate.
